src/misc/unicode.cpp

```diff
@@ -0,0 +1,1159 @@
+        if (it_alias != mappings_aliases.end())
+            mapping_aliases = &it_alias->second;
+    }
+
+    // Handle code points which are 7-bit ASCII characters
+    auto push_7bit = [&str_out](const Grapheme &grapheme) {
+        if (grapheme.HasMark())
+            return false; // not a 7-bit ASCII character
+
+        const auto code_point = grapheme.GetCodePoint();
+        if (code_point >= threshold_non_ascii)
+            return false; // not a 7-bit ASCII character
+
+        str_out.push_back(static_cast<char>(code_point));
+        return true;
+    };
+
+    // Handle code points belonging to selected code page
+    auto push_code_page = [&str_out](const code_page_mapping_t *mapping,
+                                     const Grapheme &grapheme) {
+        if (!mapping)
+            return false;
+        const auto it = mapping->find(grapheme);
+        if (it == mapping->end())
+            return false;
+
+        str_out.push_back(uint8_to_char(it->second));
+        return true;
+    };
+
+    // Handle code points which can only be mapped to ASCII
+    // using a fallback UTF8 mapping table
+    auto push_fallback = [&str_out](const Grapheme &grapheme) {
+        if (grapheme.HasMark())
+            return false;
+
+        const auto it = mapping_ascii.find(grapheme.GetCodePoint());
+        if (it == mapping_ascii.end())
+            return false;
+
+        str_out.push_back(uint8_to_char(it->second));
+        return true;
+    };
+
+    // Handle unknown code points
+    auto push_unknown = [&str_out](const uint16_t code_point) {
+        str_out.push_back(static_cast<char>(unknown_character));
+        warn_code_point(code_point);
+    };
+
+    for (size_t i = 0; i < str_in.size(); ++i) {
+        Grapheme grapheme(str_in[i]);
+        while (i + 1 < str_in.size() && is_combining_mark(str_in[i + 1])) {
+            ++i;
+            grapheme.AddMark(str_in[i]);
+        }
+
+        auto push = [&](const Grapheme &grapheme) {
+            return push_7bit(grapheme) ||
+                   push_code_page(mapping, grapheme) ||
+                   push_code_page(mapping_aliases, grapheme) ||
+                   push_fallback(grapheme);
+        };
+
+        if (push(grapheme))
+            continue;
+
+        if (grapheme.HasMark()) {
+            grapheme.StripMarks();
+            if (push(grapheme))
+                continue;
+        }
+
+        push_unknown(grapheme.GetCodePoint());
+    }
+}
+
+// ***************************************************************************
+// Read resources from files
+// ***************************************************************************
+
+static bool prepare_code_page(const uint16_t code_page);
+
+template <typename T1, typename T2>
+bool add_if_not_mapped(std::map<T1, T2> &mapping, T1 first, T2 second)
+{
+    if (mapping.count(first))
+        return false;
+
+    mapping[first] = second;
+    return true;
```

Comment:
In this case, two scans are done through the map (one to count, and second to add).

`try_emplace` can be used to do it in one pass, gives back a pair of values:

```c++
[[maybe_unused]] const auto &[item, was_added] = 
     mapping.try_emplace(first, second);
return was_added;
```

If the key already exists, `was_added` will be false.

Reply:
maybe_unused, because we don't need `item`.

Reply:
OK, changed.